Include product images in products to the get products endpoints

raisa16 opened the pull request:
-Create ImageDto to define the required attributes from PoductImage ata.

-Update ProductResponse and  ProductServiceImpl to add the list of objects with productImage endpoints and status.

raisa16 committed:
-Create ImageDto to define the required attributes from PoductImage data.
-Update ProductResponse and  ProductServiceImpl to add the list of objects with productImage endpoints and status.

raisa16 committed:
Merge branch 'develop' into feature/issue-25/include-productimages-in-products-to-get-product-images-endpoints

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/controllers/requests/ImageDto.java:
```diff
@@ -0,0 +1,37 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.nurserygarden.ecommerceapp.repositories.entities.Status;
+
+public class ImageDto {
```
Para tener más claro el nombre con que está relacionado, sería mejor incluir la palabra Product. Esto porque son las imagenes de los productos

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/services/ProductServiceImpl.java:
```diff
@@ -104,6 +109,12 @@ public void deleteById(Long id) {
 
+        List<ProductImage> productWithImages = new ArrayList<>();
+
+        productImagesRepository.findAll().forEach(productWithImages::add);
```
Esto puede tener mejor rendimiento buscando sólo las del producto.

raisa16 replied:
Actualice el ProductImageRepository para obterner la informacion necesaria por medio del productId

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/services/ProductServiceImpl.java:
```diff
@@ -104,6 +109,12 @@ public void deleteById(Long id) {
+
+        productImagesRepository.findAll().forEach(productWithImages::add);
+
+        List<ImageDto> images = productWithImages.stream().filter(image -> image.getProduct().getId().equals(product.getId())).map(image -> new ImageDto(image.getImageUrl(), image.getStatus())).collect(Collectors.toList());
```
Entiendo que se tiene la necesidad de usar caracteristicas de Java 8, pero para este caso de uso puede tener un impacto en el rendimiento del proceso. Se puede simplificar más desde base de datos. Al menos la parte del filtro de las imágenes del producto, la transformación a DTO sí es valida.

raisa16 replied:
Cree un metodo para transformar la informacion de  obtenida del ProductImage y guardarla en una lista de ProductImageDto para posteriormente agregar estos datos  a la respuesta del producto

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/services/ProductServiceImpl.java:
```diff
@@ -112,11 +123,11 @@ private ProductResponse toProductResponse(Product product) {
         productResponse.setPrice(product.getPrice());
         productResponse.setColor(product.getColor());
         productResponse.setCategoryName(product.getCategory().getName());
+        productResponse.setImageDto(images);
```
Definamos un nombre al atributo que tenga más claro que es una lista.

raisa16 replied:
decidi pasar la llamada al metodo para no crear una variable solo para pasar los  datos, sin embargo los nombres de las variables usadas en el proceso fueron actualizados para que sean mas claras

elbarbaro replied:
Sí, eso es a lo que me refería. El comentario iba más por el de lado del nombre del atributo que ya fue renombrado y necesita nuevamente el ser actualizado.

raisa16 committed:
-Update class name from ImageDto to ProductImageDto
-Update ProductImageRepository to filter by podructId
-ProductServiceImpl add toProductImageDto method to transform products to  ProductImageDto and pass sent the data to toProductResponse

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/controllers/responses/ProductResponse.java:
```diff
@@ -19,6 +21,8 @@ public class ProductResponse {
     @JsonProperty("category_name")
     private String categoryName;
+    @JsonProperty("images")
+    private List<ProductImageDto> productProductImageDtoList;
```
Para la organización de los datos está correcto el `List<ProductImageDto>` a nivel del nombre del atributo, lo podemos dejar como` images` sabiendo que es una lista de ProductImageDto

raisa16 committed:
-Update the name of attribute from productProductImageDtoList to images in ProductResponse and the call to set the atributte in ProductServiceImpl

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/services/ProductServiceImpl.java:
```diff
@@ -98,12 +101,24 @@ public void deleteById(Long id) {
         productRepository.save(product);
+    }
+
+    public List<ProductImageDto> toProductImageDto(List<ProductImage> products) {
```
Veo bien la separación de este paso en una función. Sólo que el nombre da entender que se está hablando de una una cosa en singular toProductImageDto, cuando el método donde se está pasando lo regresado por el método se llama `setImages`. Podemos cambiar al nombre a algo más claro que hable de una lista.

raisa16 replied:
actualizado a toProductImageDtoList

raisa16 committed:
-Update the name of method toProductImagedto now is going to be called toProductResponse
